### PR TITLE
feat: add not found component

### PR DESCRIPTION
**Pull request by JonathanXu1**
## Description
This is my submission to the GCI task for designing a 404 landing page. It also utilizes particles.js for an interactive background. To do this, I created a NotFound component in src/app. 
Updated from original pull request: https://github.com/openMF/web-app/pull/326

## Related issues and discussion
#110 

## Screenshots, if any
image

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

File: angular.json [hunk 1]
```diff
@@ -62,7 +62,9 @@
                 "lazy": true
               }
             ],
-            "scripts": []
+            "scripts": [
```
Comment (abhaychawla) [inline angular.json]: This is not required as you included the script via a cdn in `index.html` (which is a better way). Remove this change (file) from your commit.

File: package.json [hunk 2]
```diff
@@ -46,6 +46,7 @@
     "core-js": "2.5.7",
     "hammerjs": "2.0.8",
     "lodash": "4.17.10",
+    "particles.js": "^2.0.0",
```
Comment (abhaychawla) [inline package.json]: This is not required as you included the script via a cdn in `index.html` (which is a better way). Remove this change (file) from your commit.

File: src/app/app-routing.module.ts [hunk 3]
```diff
@@ -2,14 +2,20 @@
...
 const routes: Routes = [
+  {
+    path: '404',
+    component: NotFoundComponent
+  },
   {
     path: '**',
-    redirectTo: '',
-    pathMatch: 'full'
+    redirectTo: '/404'
```
Comment (abhaychawla) [inline src/app/app-routing.module.ts]: As I told you earlier, remove the redirect route and just simply add a new path for 404. Update the code with this:
```typescript
{
    path: '**',
    component: PageNotFoundComponent
}
```

File: src/app/not-found/not-found.component.html [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        This page requires JavaScript to work properly. Please enable JavaScript in your browser.
+      </p>
+    </noscript>
+
+    <script>
+      // This hides the address bar on mobile browsers
+      // https://developers.google.com/web/fundamentals/native-hardware/fullscreen/
+      window.addEventListener('load',function() { window.scrollTo(0, 1); });
+    </script>
+
```
Comment (abhaychawla) [inline src/app/not-found/not-found.component.html]: Remove the code from lines 1-39 and 45-47, a component only contains part of html code, which is loaded in place of the `<mifosx-web-app>` selector present in `index.html`, so these parts are not supposed to be present here.

File: src/app/not-found/not-found.component.scss [hunk 5]
```diff
@@ -0,0 +1,33 @@
+canvas {
```
Comment (abhaychawla) [inline src/app/not-found/not-found.component.scss]: Remove this scss. Its not being used anywhere.

File: src/app/not-found/not-found.component.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  templateUrl: './not-found.component.html',
+  styleUrls: ['./not-found.component.scss']
+})
+export class NotFoundComponent implements OnInit {
+
+  constructor() { }
+
+  ngOnInit() {
+    particlesJS.load('particles-js', '/assets/data/particles.json', function() {
+      console.log('particles loaded');
```
Comment (abhaychawla) [inline src/app/not-found/not-found.component.ts]: Remove the callback function and console log statement.

File: src/app/not-found/not-found.component.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  selector: 'mifosx-not-found',
+  templateUrl: './not-found.component.html',
+  styleUrls: ['./not-found.component.scss']
+})
+export class NotFoundComponent implements OnInit {
+
+  constructor() { }
+
+  ngOnInit() {
+    particlesJS.load('particles-js', '/assets/data/particles.json', function() {
```
Comment (abhaychawla) [inline src/app/not-found/not-found.component.ts]: Add the file into the assets folder directly, no need to create a new folder data to store it.

File: src/index.html [hunk 8]
```diff
@@ -36,6 +36,9 @@
     window.addEventListener('load',function() { window.scrollTo(0, 1); });
   </script>
 
+  <!-- ParticlesJS -->
```
Comment (abhaychawla) [inline src/index.html]: Place this script import just before the body tag is closed, not in between.

File: src/app/app-routing.module.ts (hunk 3 above)
Comment (JonathanXu1) [reply src/app/app-routing.module.ts]: Sorry I believe I didn't reapply the changes when I squashed my commits to the wrong version.

File: src/app/not-found/not-found.component.html [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    <link rel="icon" type="image/png" href="assets/images/MifosX_logo.png" sizes="182x184">
+    <link rel="apple-touch-icon" type="image/png" href="apple-touch-icon.png" sizes="182x184"/>
+    <link rel="manifest" href="manifest.json"/>
+
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+    <link href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Mono:300" rel="stylesheet">
+    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
+  </head>
+  
+  <body>
```
Comment (abhaychawla) [inline src/app/not-found/not-found.component.html]: Jonathan, as I told you earlier remove the code from lines 1-21 and 27-28, a component only contains part of html code, which is loaded in place of the `<mifosx-web-app>` selector present in `index.html`, so these parts are not supposed to be present here. These parts are already there in `index.html`.

File: src/app/not-found/not-found.component.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  selector: 'mifosx-not-found',
+  templateUrl: './not-found.component.html',
+  styleUrls: ['./not-found.component.scss']
+})
+export class NotFoundComponent implements OnInit {
+
+  constructor() { }
+
+  ngOnInit() {
+    particlesJS.load('particles-js', '/assets/particles.json', null);
```
Comment (abhaychawla) [inline src/app/not-found/not-found.component.ts]: Do not pass a null argument, you can simply right it as `particlesJS.load('particles-js', '/assets/particles.json');` without any argument. Its not necessary and that should work.

**Comment by JonathanXu1** [comment]
Sorry, I've removed the commit messages, the null argument and cleaned up the not-found.component.html file.

File: src/app/not-found/not-found.component.html [hunk 11]
```diff
@@ -0,0 +1,10 @@
+<!doctype html>
+<html lang="en" style="height: 100%">
+  <body>
```
Comment (abhaychawla) [inline src/app/not-found/not-found.component.html]: The file still contains lines 1-3 and 9-10 which are not required.

File: src/assets/particles.json [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+      "push": {
+        "particles_nb": 4
+      },
+      "remove": {
+        "particles_nb": 2
+      }
+    }
+  },
+  "retina_detect": true
+}
```
Comment (abhaychawla) [inline src/assets/particles.json]: Add a new line at the end of file to avoid this.

**Pull request merged**